Documentation/design-docs/PinnedHeap.md

```diff
@@ -0,0 +1,160 @@
+
+In order to group pinned objects together we have a couple of options –
+
+1) We allow pinning at allocation time so when such an object is allocated we put it on the pinned heap. 
+
+2) We allow migrating non pinned objects to the pinned heap at the time they need to be pinned.
+
+1) is a much cleaner solution. Of course it also imposes a limitation which is the user needs to have control over the objects they want to pin and is ok with pinning it at allocation time. This is not always the case. However, we believe that this feature is the most important for framework components which often take care of the underlying pinning and they are in general ok with this limitation so we will go with option 1). If we ever decide that it’s important to do 2) we can add it in the future.
+
+An object allocated on the pinned heap can be referenced by other objects normally. It’s pinned by the virtue of being on the pinned heap so we do not distinguish between a pinned reference (like a pinned GC handle) and a non pinned reference. This means as long as the object is still referenced, it will stay on the pinned heap and will not be moved.
+
+## API - allocating an array on the pinned heap
+
+For users who want to allocate their objects pinned, we provide [a new API](https://github.com/dotnet/corefx/issues/31787) to allocate such an object. The API declaration is the following:
+
+```csharp
+class GC
+{
+    // generation: -1 means to let GC decide (equivalent to regular new T[])
+    // 0 means to allocate in gen0
+    // GC.MaxGeneration means to allocate in the oldest generation
+    //
+    // pinned: true means you want to pin this object that you are allocating
+    // otherwise it's not pinned.
+    //
+    // alignment: only supported if pinned is true.
+    // -1 means you don't care which means it'll use the default alignment.
+    // otherwise specify a power of 2 value that's >= pointer size
+    // the beginning of the payload of the object (&array[0]) will be aligned with 
+    // this alignment.
+    static T[] AllocateArray<T>(int length, 
+                                int generation = -1, 
+                                bool pinned = false, 
+                                int alignment = -1);
+}
+```
+
+Note that for pinned heap we are only implementing the relevant part of the API, meaning only 2 arguments will be interpreted – length and pinned. Since this is a user facing feature we impose the following limitations on the objects allocated by this API which will make the implementation cleaner and the perf better while not hindering the majority of the usage:
```

Comment:
> we are only implementing the relevant part of the API, meaning only 2 arguments will be interpreted – length and pinned

Why isn't "alignment" relevant?

Reply:
it's only not relevant in the sense of we are not going to implement the alignment for V1 of this feature. as I mentioned at the end of the doc it's something we'll be thinking about for future development.

Reply:
> it's only not relevant in the sense of we are not going to implement the alignment for V1 of this feature

Ok.  I realize this is just a design document for the approach and the API here is a sketch.  That said, we shouldn't expose in an API functionality that we haven't implemented and won't until some undefined point in the future.  That could mean we start with an overload that doesn't contain those unimplemented arguments and then add an overload that does later when they're implemented, or we could have a config struct with these as properties that we add more to later, or some such thing.

Reply:
that's fine by me. of course we'll goto the API review for this and get the final decl of the API figured out.

Reply:
Sounds good.